WarehouseManagement.Core/Services/VendorService.cs

```diff
@@ -0,0 +1,232 @@
+
+            vendor.Name = model.Name;
+            vendor.SystemNumber = model.SystemNumber;
+            vendor.LastModifiedAt = DateTime.UtcNow;
+            vendor.LastModifiedByUserId = userId;
+
+            await repository.SaveChangesWithLogAsync();
+        }
+
+        public async Task DeleteAsync(int id, string userId)
+        {
+            var vendor = await repository
+                .All<Vendor>()
+                .Where(v => v.Id == id)
+                .Include(v => v.Deliveries)
+                .FirstOrDefaultAsync();
+
+            if (vendor == null)
+            {
+                throw new KeyNotFoundException($"Vendor with ID {id} not found.");
+            }
+
+            if (vendor.Deliveries.Any())
+            {
+                var deliveries = vendor.Deliveries.Select(d => d.SystemNumber).ToList();
+                throw new InvalidOperationException(
+                    $"Vendor can not be deleted because has existing deliveries with system numbers: {string.Join(", ", deliveries)}"
+                );
+            }
+
+            vendor.DeletedByUserId = userId;
+
+            repository.SoftDelete(vendor);
+
+            await repository.SaveChangesAsync();
+        }
+
+        public async Task<string> RestoreAsync(int id)
+        {
+            var vendor = await repository.All<Vendor>().FirstOrDefaultAsync(v => v.Id == id);
+
+            if (vendor == null)
+            {
+                throw new KeyNotFoundException($"Vendor with ID {id} not found.");
+            }
+
+            if (!vendor.IsDeleted)
+            {
+                throw new InvalidOperationException($"Vendor {vendor.Name} is not deleted");
+            }
+
+            if (await this.AnotherVendorWithNameExistAsync(vendor.Id, vendor.Name))
+            {
+                throw new InvalidOperationException(
+                    $"Another vendor with name {vendor.Name} already exist"
+                );
+            }
+
+            if (await this.AnotherVendorWithSystemNumberExistAsync(vendor.Id, vendor.SystemNumber))
+            {
+                throw new InvalidOperationException(
+                    $"Another vendor with system number {vendor.SystemNumber} already exist"
+                );
+            }
+
+            repository.UnDelete(vendor);
+            await repository.SaveChangesAsync();
+
+            return vendor.Name;
+        }
+
+        public async Task<bool> ExistByNameAsync(string name)
+        {
+            return await repository
+                .AllReadOnly<Vendor>()
+                .AnyAsync(v => v.Name.ToLower() == name.ToLower());
+        }
+
+        public async Task<bool> ExistBySystemNumberAsync(string systemNumber)
+        {
+            return await repository
+                .AllReadOnly<Vendor>()
+                .AnyAsync(v => v.SystemNumber.ToLower() == systemNumber.ToLower());
+        }
+
+        public async Task<bool> AnotherVendorWithNameExistAsync(int id, string name)
```

Comment:
whats exactly the difference between `ExistsByNameAsync` and `AnotherVendorWithNameExistsAsync`? Looks like they are the same, no?

Reply:
```c#
  public async Task<bool> AnotherVendorWithNameExistAsync(int id, string name)
  {
      return await repository
          .AllReadOnly<Vendor>()
          .Where(v => v.Id != id)
          .AnyAsync(v => v.Name.ToLower() == name.ToLower());
  }
```
The `AnotherVendorWithNameExistsAsync` method is used when an edit request is sent. This service checks if another entity has the same name, excluding the one currently being edited. I cannot use the `VendorWithNameExistsAsync` method because it checks all entities. If the client only wants to edit the system number but not the name, `VendorWithNameExistsAsync` will return "true," indicating that a vendor with that name already exists. This check can be implemented in the `EditAsync` service method if needed, allowing for the removal of `AnotherVendorWithNameExistsAsync`.